Add yaml output

cbm755 opened the pull request:
Fixes #37 

### todo
- [x] add test
- [x] add another test for the req file
- [x] ~~`build_module_as_str` seems unused and untested, same with the new `build_module_as_yaml_str`.~~  Filed #45 for this.
- [x] local imports for `pyyaml` dep, not at top
- [x] use file-name to guess output (default json) 
- [x] add `--yaml` flag
- [x] `--platform-info` should respect `--yaml` flag

### help wanted
- [x] poetry extra dep (I'm not a poetry user, not sure about this
- [x] left a TODO

cbm755 committed:
Consistent json indent

cbm755 committed:
Write yaml if output file ends in yaml or yml extension

Start of a fix for Issue #37.

real-yfprojects commented on req2flatpak.py:
```diff
@@ -682,6 +695,10 @@ def cli_parser() -> argparse.ArgumentParser:
         nargs="?",
         type=argparse.FileType("w"),
         default=sys.stdout,
+        help="""
+            By default, writes JSON but specify a '.yaml' extension and YAML
+            will be written instead, provided you have the 'pyyaml' package.
+        """,
```
This is a nice mechanism. I guess a `--yaml` option would be useful nonetheless. For example for writing to stdout.

real-yfprojects commented on pyproject.toml:
```diff
@@ -19,6 +19,7 @@ packaging = { version = "^21.3", optional = true }
 
 [tool.poetry.extras]
 packaging = ["packaging"]
+# TODO: something about pyyaml here
```
```suggestion
yaml = ["pyyaml"]
```

real-yfprojects commented on req2flatpak.py:
```diff
@@ -634,12 +634,25 @@ def sources(downloads: Iterable[Download]) -> List[dict]:
+        # optional dependency, not imported at top
+        import yaml
+
+        return yaml.dump(cls.build_module(*args, **kwargs), indent=2)
```
Maybe setting `default_flow_style` to force block format?

cbm755 committed:
Update pyproject.toml

Co-authored-by: yfprojects <62463991+real-yfprojects@users.noreply.github.com>

cbm755 committed:
Add a "--yaml" cmdline arg

cbm755 committed:
write yaml from '--platform-info' too

real-yfprojects commented on req2flatpak.py:
```diff
@@ -705,10 +728,31 @@ def main():
     parser = cli_parser()
     options = parser.parse_args()
 
+    def _get_yaml_module_or_exit():
+        try:
+            # optional dependency, not imported at top
+            import yaml
+        except ImportError:
+            parser.error(
+                "Outputing YAML requires 'pyyaml' package: try 'pip install pyyaml'"
+        )
+        return yaml
+
```
I don't oppose adding this function although I did wonder why you don't add another if-statement:

```py
if options.yaml:
    try:
        # optional dependency, not imported at top
        import yaml
    except ImportError:
        parser.error(
            "Outputing YAML requires 'pyyaml' package: try 'pip install pyyaml'"
    )

cbm755 replied:
seemed like we need this functionality in two places (and maybe three, although I'm not sure how to wire it into the classmethod).

But sure I can inline it.

cbm755 replied:
nm, I see another way

real-yfprojects replied:
> seemed like we need this functionality in two places

When using `parser` the function doesn't bring much value since that can be achieved with the single if-statement too.

> I'm not sure how to wire it into the classmethod

I suggest that the `FlatpakGenerator` API just forwards the ImportError, maybe with a more descriptive error message.

real-yfprojects commented on req2flatpak.py:
```diff
@@ -705,10 +728,31 @@ def main():
+        output_stream = sys.stdout
 
     # print platform info if requested, and exit
+    if options.platform_info and options.yaml:
+        yaml = _get_yaml_module_or_exit()
+        yaml.dump(
+            asdict(PlatformFactory.from_current_interpreter()), output_stream, indent=2
+        )
+        parser.exit()
```
Not sure whether to value code simplicity over redundancy. I would prioritize the latter and use a nested if-statement.

cbm755 committed:
Refactor yaml

real-yfprojects commented:
Are you using [pre-commit](pre-commit.com)?

cbm755 commented:
> Are you using pre-commit

~~No I have never learned that, was always worried it would mess up by other projects.  I do manually run black before committing, but I understand maybe there are other standards at play here...~~

**update** luddite-mode turned off, trying it...

cbm755 committed:
Force default flow style to False

I think this is for older pyyaml, AFAICT, False is the default circa
2022.  No need to specific indent=2 as that seems to be the default.

cbm755 commented:
pre-commit mucked around in `poetry.lock` with lots of unrelated changes.  I'll not commit those.

`pylama` gaves lots of strange-looking errors so I uninstalled that.


```
poetry-check.............................................................Passed
poetry-lock..............................................................Passed
Sync with Poetry.........................................................Passed
isort (python)...........................................................Passed
black....................................................................Passed
prettier.................................................................Passed
rstcheck.................................................................Passed
pylama...................................................................Failed
- hook id: pylama
- exit code: 1

Executable `pylama` not found

bandit...................................................................Failed
- hook id: bandit
- exit code: 1

[node_visitor]	WARNING	Unable to find qualified name for module: req2flatpak.py
--------------------------------------------------
 tests/test_req2flatpak.py:106:
 [  B506  ] Severity: MEDIUM Confidence: HIGH
 Use of unsafe yaml load. Allows instantiation of arbitrary objects. Consider yaml.safe_load().
```

The yaml.safe_load() seems nothing to worry about b/c its in a test.

cbm755 commented:
now I can't commit, not too impressed so far...

```
[WARNING] Stashed changes conflicted with hook auto-fixes... Rolling back fixes...
[INFO] Restored changes from /home/cbm/.cache/pre-commit/patch1677370833-285058.
```

cbm755 commented:
... and installing it does not go well:

```
git commit
/usr/bin/python3: No module named pre_commit
```

So back to luddite mode for me I think!

cbm755 committed:
Add test for yaml output

cbm755 commented:
I'm not very confident about the test: please take a look at that.

real-yfprojects commented:
> So back to luddite mode for me I think!

Don't blame me, but I had to smile reading about your struggle. Our pre-commit setup is not prefect yet. I tweaked it in #42 and #39.

> No I have never learned that, was always worried it would mess up by other projects.

You can instal it on a pre-repository basis.

real-yfprojects commented on tests/test_req2flatpak.py:
```diff
@@ -95,6 +97,15 @@ def test_cli_with_reqs_as_args(self):
+        args += ["--target-platforms"] + self.target_platforms
+        args += ["--yaml"]
+        result = self._run_r2f(args)
+        build_module = yaml.load(result.stdout, yaml.Loader)
```
I don't think specifying `loader` is needed. The rest looks good though.

cbm755 replied:
Without that, I get:
```
self = <tests.test_req2flatpak.EggTest testMethod=test_cli_with_reqs_as_args_yaml>

    def test_cli_with_reqs_as_args_yaml(self):
        """Runs req2flatpak in yaml mode by passing requirements as cmdline arg."""
        args = ["--requirements"] + self.requirements
        args += ["--target-platforms"] + self.target_platforms
        args += ["--yaml"]
        result = self._run_r2f(args)
>       build_module = yaml.load(result.stdout)
E       TypeError: load() missing 1 required positional argument: 'Loader'
```

Note: I'm not 100% sure I'm passing the correct thing, it does seem like this would be the sensible default...

real-yfprojects replied:
You are right, I had a look into the codebase. However [the docs](https://pyyaml.org/wiki/PyYAMLDocumentation#cb67) say something else.

real-yfprojects replied:
I found these yaml/pyyaml#576 and yaml/pyyaml.org#15. A comment in the latter suggests using `safe_load` since that doesn't need a loader.

real-yfprojects commented:
To fix the ci you will have to run the following:

```sh
poetry lock --no-update && SKIP=poetry-lock git commit
```

cbm755 committed:
Add another yaml test for requirements file

cbm755 committed:
test: use yaml safe_load which is simpler

Also avoids warnings from linters.

cbm755 commented:
This is ready to go from my point-of-view, but maybe needs some clean-up about testing with optional dependency `pyyaml` (which is not optional for testing).

real-yfprojects commented:
I can add a commit for fixing ci.

real-yfprojects commented on tests/test_req2flatpak.py:
```diff
@@ -107,6 +118,19 @@ def test_cli_with_reqs_as_file(self):
 
+    def test_cli_with_reqs_as_file_yaml(self):
+        """Runs req2flatpak by passing requirements as requirements.txt file."""
+        with tempfile.NamedTemporaryFile(mode="w+", encoding="utf-8") as req_file:
+            req_file.write("\n".join(self.requirements))
+            req_file.flush()
+            req_file.seek(0)
+            args = ["--requirements-file", req_file.name]
```
Maybe implement a common context manager:
```python
from contextlib import contextmanager

@contextmanager
def requirements_file():
    with tempfile.NamedTemporaryFile(mode="w+", encoding="utf-8") as req_file:
            req_file.write("\n".join(self.requirements))
            req_file.flush()
            req_file.seek(0)
            yield req_file
        
with requirements_file() as req_file:
    args = ["--requirements-file", req_file.name]
```

johannesjh replied:
nice!

real-yfprojects committed:
Fix extra dependency configuration and add pyyaml to lock file.

* pyproject.toml (tool.poetry.dependencies): Add `pyyaml` as optional dependency.

* poetry.lock : Run `poetry lock --no-update` to add pyyaml.

johannesjh commented on req2flatpak.py:
```diff
@@ -771,6 +810,10 @@ def main():
+        parser.exit()
+
     # write output
     json.dump(build_module, output_stream, indent=4)
```
Could you fix #45 by calling `build_module_as_str` instead of calling `json.dump` directly, thx

johannesjh commented on req2flatpak.py:
```diff
@@ -771,6 +810,10 @@ def main():
     build_module = FlatpakGenerator.build_module(requirements, downloads)
 
+    if options.yaml:
+        yaml.dump(build_module, output_stream, default_flow_style=False)
```
Could you call `build_module_as_yaml_str` instead of `yaml.dump` please

real-yfprojects committed:
Add `types-pyyaml` to lint dev packages.

* pyproject.toml
* poetry.lock

real-yfprojects committed:
Merge branch 'main' into pr/cbm755/41

real-yfprojects committed:
Run ci tests with multiple sets of extra dependencies.

This adds a matrix that runs the tests once with `packaging`, `yaml` and both.

* .github/workflows/ci.yml

* tests/test_req2flatpak.py : Skip yaml tests when pyyaml isn't installed.

real-yfprojects committed:
Generate test matrix dynamically for extras.

* .github/workflows/ci.yml

real-yfprojects commented on tests/test_req2flatpak.py:
```diff
@@ -7,8 +7,14 @@
+from unittest import skipUnless
 from unittest.mock import patch
 
+try:
+    import yaml
+except ImportError:
+    yaml = None  # type: ignore
```
Not sure whether there is a more elegant solution.

johannesjh replied:
looks ok to me, easy to understand

johannesjh commented:
hi guys, this PR is almost finished, well done! we have a few open comments from the code review and three complaints from pylint. @cbm755 are you going to walk the last mile? I hope you are not getting annoyed, do you need / want help?

real-yfprojects commented:
> three complaints from pylint

I don't know about the complexity complaint. The import complaints could be resolved similar to how I did it in the tests.

johannesjh commented:
I would suggest to ignore the complexity constraint. 
* The complaint is [pylint R0912 "too many branches"](https://pylint.pycqa.org/en/latest/user_guide/messages/refactor/too-many-branches.html). 
* A proper solution would be to refactor the code and split the main methods into multiple methods. I doubt this will make the code easier to understand or maintain.
* Let's ignore the complaint and work around the issue by commenting the line with `# pylint: disable=too-many-branches`.

real-yfprojects committed:
Move import of `yaml` to the beginning of `req2flatpak.py`.

Yaml is tried to be imported else `yaml` is set to `None`. Later the code will only check wether `yaml` is True (is not None).

* req2flatpak.py

real-yfprojects committed:
Ignore too-many-branches in `main` function.

* req2flatpak.py : Ignore pylint complain `too-many-branches` for `main`.

real-yfprojects committed:
Use a common contextmanager to create requirements file in tests.

* tests/test_req2flatpak.py (Req2FlatpakBaseTest.requirements_file): Implement this method returning a contextmanager
	that creates and prefills a requirements file with requirements.

* tests/test_req2flatpak.py (Req2FlatpakBaseTest.test_cli_with_reqs_as_file[_yaml]): Use `requirements_file()`.

real-yfprojects committed:
Use `FlatpakGenerator.build_module_as_[yaml_]str` in `main`.

These methods implement formatting, so why duplicate that code. Fixes #45.

* req2flatpak.py (main)

cbm755 commented:
Thanks @real-yfprojects for the edits.  I tested this and it works for me!

I do notice that the yaml dict out is in a rather unpleasant order:
```
build-commands:
- pip3 install --verbose --exists-action=i --no-index --find-links="file://${PWD}"
  --prefix=${FLATPAK_DEST} --no-build-isolation pymupdf
buildsystem: simple
name: python3-package-installation
sources:
- only-arches:
  - aarch64
  sha256: 36181c4cb27740791d611d0224dd18ac78e23a1f06aa2151394c5b9194b4f885
  type: file
  url: https://files.pythonhosted.org/packages/9c/90/4dcdb731e9aaecfa2f10300bf51f4f2d2079083819fd4a109a3e1d331338/PyMuPDF-1.21.1-cp310-cp310-manylinux_2_17_aarch64.manylinux2014_aarch64.whl
- only-arches:
  - x86_64
  sha256: e5273ff0c3bf08428ef956c4a5e5e0b475667cc3e1cb7b41d9f5131636996e59
  type: file
  url: https://files.pythonhosted.org/packages/e0/27/83c843ef54f5c65874c844a77e216326eb5b2a2fcb08bd7cb0784fb759f9/PyMuPDF-1.21.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
```
Note "build-commands" before "name" for example.  And "sha256" before "type".  Its still valid but I wonder why that happens?  Should I file an new issue?

For comparison, here is the JSON:
```
{
    "name": "python3-package-installation",
    "buildsystem": "simple",
    "build-commands": [
        "pip3 install --verbose --exists-action=i --no-index --find-links=\"file://${PWD}\" --prefix=${FLATPAK_DEST} --no-build-isolation pymupdf"
    ],
    "sources": [
        {
            "type": "file",
            "url": "https://files.pythonhosted.org/packages/9c/90/4dcdb731e9aaecfa2f10300bf51f4f2d2079083819fd4a109a3e1d331338/PyMuPDF-1.21.1-cp310-cp310-manylinux_2_17_aarch64.manylinux2014_aarch64.whl",
            "sha256": "36181c4cb27740791d611d0224dd18ac78e23a1f06aa2151394c5b9194b4f885",
            "only-arches": [
                "aarch64"
            ]
        },
        {
            "type": "file",
            "url": "https://files.pythonhosted.org/packages/e0/27/83c843ef54f5c65874c844a77e216326eb5b2a2fcb08bd7cb0784fb759f9/PyMuPDF-1.21.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl",
            "sha256": "e5273ff0c3bf08428ef956c4a5e5e0b475667cc3e1cb7b41d9f5131636996e59",
            "only-arches": [
                "x86_64"
            ]
        }
    ]
}                         
```
My understanding of modern Python is that dicts have a consistent unofficial ordering based on the the order you insert things into them.  In this case, that translates into a pleasant ordering when outputing JSON but not YAML...

cbm755 committed:
Fix dict-ordering when outputing yaml

cbm755 commented:
I found a very long thread about it [here](https://github.com/yaml/pyyaml/issues/110) where I discovered there is a `sort_keys=False` option.

real-yfprojects commented:
> I do notice that the yaml dict out is in a rather unpleasant order:

Good call!

johannesjh commented:
well done, thank you for the effort! ready to merge. 🎉
